doxygen/man/man3/SendCoinsDialog.3
.TH "SendCoinsDialog" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SendCoinsDialog \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <sendcoinsdialog\&.h>\fP
.PP
Inherits QDialog\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBclear\fP ()"
.br
.ti -1c
.RI "void \fBreject\fP ()"
.br
.ti -1c
.RI "void \fBaccept\fP ()"
.br
.ti -1c
.RI "\fBSendCoinsEntry\fP * \fBaddEntry\fP ()"
.br
.ti -1c
.RI "void \fBupdateRemoveEnabled\fP ()"
.br
.ti -1c
.RI "void \fBsetBalance\fP (const \fBCAmount\fP &balance, const \fBCAmount\fP &stake, const \fBCAmount\fP &unconfirmedBalance, const \fBCAmount\fP &immatureBalance, const \fBCAmount\fP &anonymizedBalance, const \fBCAmount\fP &watchOnlyBalance, const \fBCAmount\fP &watchOnlyStake, const \fBCAmount\fP &watchUnconfBalance, const \fBCAmount\fP &watchImmatureBalance)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSendCoinsDialog\fP (QWidget *parent=0)"
.br
.ti -1c
.RI "\fB~SendCoinsDialog\fP ()"
.br
.ti -1c
.RI "void \fBsetModel\fP (\fBWalletModel\fP *model)"
.br
.ti -1c
.RI "QWidget * \fBsetupTabChain\fP (QWidget *\fBprev\fP)"
.br
.ti -1c
.RI "void \fBpasteEntry\fP (const \fBSendCoinsRecipient\fP &rv)"
.br
.ti -1c
.RI "bool \fBhandleURI\fP (const QString &uri)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Dialog for sending darksilks 
.PP
Definition at line 23 of file sendcoinsdialog\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SendCoinsDialog::SendCoinsDialog (QWidget * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 25 of file sendcoinsdialog\&.cpp\&.
.SS "SendCoinsDialog::~SendCoinsDialog ()"

.PP
Definition at line 141 of file sendcoinsdialog\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void SendCoinsDialog::accept ()\fC [slot]\fP"

.PP
Definition at line 310 of file sendcoinsdialog\&.cpp\&.
.SS "\fBSendCoinsEntry\fP * SendCoinsDialog::addEntry ()\fC [slot]\fP"

.PP
Definition at line 315 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::clear ()\fC [slot]\fP"

.PP
Definition at line 291 of file sendcoinsdialog\&.cpp\&.
.SS "bool SendCoinsDialog::handleURI (const QString & uri)"

.PP
Definition at line 396 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::pasteEntry (const \fBSendCoinsRecipient\fP & rv)"

.PP
Definition at line 373 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::reject ()\fC [slot]\fP"

.PP
Definition at line 305 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::setBalance (const \fBCAmount\fP & balance, const \fBCAmount\fP & stake, const \fBCAmount\fP & unconfirmedBalance, const \fBCAmount\fP & immatureBalance, const \fBCAmount\fP & anonymizedBalance, const \fBCAmount\fP & watchOnlyBalance, const \fBCAmount\fP & watchOnlyStake, const \fBCAmount\fP & watchUnconfBalance, const \fBCAmount\fP & watchImmatureBalance)\fC [slot]\fP"

.PP
Definition at line 412 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::setModel (\fBWalletModel\fP * model)"

.PP
Definition at line 111 of file sendcoinsdialog\&.cpp\&.
.SS "QWidget * SendCoinsDialog::setupTabChain (QWidget * prev)"
Set up the tab chain manually, as Qt messes up the tab chain by default in some cases (issue https://bugreports.qt-project.org/browse/QTBUG-10907)\&. 
.PP
Definition at line 358 of file sendcoinsdialog\&.cpp\&.
.SS "void SendCoinsDialog::updateRemoveEnabled ()\fC [slot]\fP"

.PP
Definition at line 336 of file sendcoinsdialog\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes

src/qt/sendcoinsdialog.cpp
#include "sendcoinsdialog.h"
#include "ui_sendcoinsdialog.h"
#include "net.h"
#include "protocol.h"
#include "chainparams.h"
#include "walletmodel.h"
#include "addresstablemodel.h"
#include "addressbookpage.h"

#include "darksilkunits.h"
#include "addressbookpage.h"
#include "optionsmodel.h"
#include "sendcoinsentry.h"
#include "guiutil.h"
#include "askpassphrasedialog.h"

#include "base58.h"
#include "coincontrol.h"
#include "coincontroldialog.h"

#include <QMessageBox>
#include <QTextDocument>
#include <QScrollBar>
#include <QSettings>
#include <QClipboard>

SendCoinsDialog::SendCoinsDialog(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::SendCoinsDialog),
    model(0),
    fNewRecipientAllowed(true)
{
    ui->setupUi(this);

#ifdef Q_OS_MAC // Icons on push buttons are very uncommon on Mac
    ui->addButton->setIcon(QIcon());
    ui->clearButton->setIcon(QIcon());
    ui->sendButton->setIcon(QIcon());
#endif

#if QT_VERSION >= 0x040700
    /* Do not move this to the XML file, Qt before 4.7 will choke on it */
    ui->lineEditCoinControlChange->setPlaceholderText(tr("Enter a DarkSilk address (e.g. D7fjE4R4r2RoEdqYk3QsAqFckyf9pRHN6i))"));
#endif

    addEntry();

    connect(ui->addButton, SIGNAL(clicked()), this, SLOT(addEntry()));
    connect(ui->clearButton, SIGNAL(clicked()), this, SLOT(clear()));

    // Coin Control
    ui->lineEditCoinControlChange->setFont(GUIUtil::darksilkAddressFont());
    connect(ui->pushButtonCoinControl, SIGNAL(clicked()), this, SLOT(coinControlButtonClicked()));
    connect(ui->checkBoxCoinControlChange, SIGNAL(stateChanged(int)), this, SLOT(coinControlChangeChecked(int)));
    connect(ui->lineEditCoinControlChange, SIGNAL(textEdited(const QString &)), this, SLOT(coinControlChangeEdited(const QString &)));
    
    // DarkSilk specific
    QSettings settings;
    if (!settings.contains("bUseSandStorm"))
        settings.setValue("bUseSandStorm", false);
    if (!settings.contains("bUseInstantX"))
        settings.setValue("bUseInstantX", false);
        
    bool useSandStorm = settings.value("bUseSandStorm").toBool();
    bool useInstantX = settings.value("bUseInstantX").toBool();

    if(fLiteMode) {
        ui->checkUseSandstorm->setChecked(false);
        ui->checkUseSandstorm->setVisible(false);
        ui->checkInstantX->setVisible(false);
        CoinControlDialog::coinControl->useSandStorm = false;
        CoinControlDialog::coinControl->useInstantX = false;
    }
    else{
        ui->checkUseSandstorm->setChecked(useSandStorm);
        ui->checkInstantX->setChecked(useInstantX);
        CoinControlDialog::coinControl->useSandStorm = useSandStorm;
        CoinControlDialog::coinControl->useInstantX = useInstantX;
    }

    connect(ui->checkUseSandstorm, SIGNAL(stateChanged ( int )), this, SLOT(updateDisplayUnit()));
    connect(ui->checkInstantX, SIGNAL(stateChanged ( int )), this, SLOT(updateInstantX()));

    // Coin Control: clipboard actions
    QAction *clipboardQuantityAction = new QAction(tr("Copy quantity"), this);
    QAction *clipboardAmountAction = new QAction(tr("Copy amount"), this);
    QAction *clipboardFeeAction = new QAction(tr("Copy fee"), this);
    QAction *clipboardAfterFeeAction = new QAction(tr("Copy after fee"), this);
    QAction *clipboardBytesAction = new QAction(tr("Copy bytes"), this);
    QAction *clipboardPriorityAction = new QAction(tr("Copy priority"), this);
    QAction *clipboardLowOutputAction = new QAction(tr("Copy dust"), this);
    QAction *clipboardChangeAction = new QAction(tr("Copy change"), this);
    connect(clipboardQuantityAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardQuantity()));
    connect(clipboardAmountAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardAmount()));
    connect(clipboardFeeAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardFee()));
    connect(clipboardAfterFeeAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardAfterFee()));
    connect(clipboardBytesAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardBytes()));
    connect(clipboardPriorityAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardPriority()));
    connect(clipboardLowOutputAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardLowOutput()));
    connect(clipboardChangeAction, SIGNAL(triggered()), this, SLOT(coinControlClipboardChange()));
    ui->labelCoinControlQuantity->addAction(clipboardQuantityAction);
    ui->labelCoinControlAmount->addAction(clipboardAmountAction);
    ui->labelCoinControlFee->addAction(clipboardFeeAction);
    ui->labelCoinControlAfterFee->addAction(clipboardAfterFeeAction);
    ui->labelCoinControlBytes->addAction(clipboardBytesAction);
    ui->labelCoinControlPriority->addAction(clipboardPriorityAction);
    ui->labelCoinControlLowOutput->addAction(clipboardLowOutputAction);
    ui->labelCoinControlChange->addAction(clipboardChangeAction);

    fNewRecipientAllowed = true;
}

void SendCoinsDialog::setModel(WalletModel *model)
{
    this->model = model;

    if(model && model->getOptionsModel())
    {
        for(int i = 0; i < ui->entries->count(); ++i)
        {
            SendCoinsEntry *entry = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(i)->widget());
            if(entry)
            {
                entry->setModel(model);
            }
        }

        setBalance(model->getBalance(), model->getStake(), model->getUnconfirmedBalance(), model->getImmatureBalance(), model->getAnonymizedBalance(),
            model->getWatchBalance(), model->getWatchStake(), model->getWatchUnconfirmedBalance(), model->getWatchImmatureBalance());
        connect(model, SIGNAL(balanceChanged(CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount)), this, SLOT(setBalance(CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount, CAmount)));
        connect(model->getOptionsModel(), SIGNAL(displayUnitChanged(int)), this, SLOT(updateDisplayUnit()));
        updateDisplayUnit();
        
        // Coin Control
        connect(model->getOptionsModel(), SIGNAL(displayUnitChanged(int)), this, SLOT(coinControlUpdateLabels()));
        connect(model->getOptionsModel(), SIGNAL(coinControlFeaturesChanged(bool)), this, SLOT(coinControlFeatureChanged(bool)));
        connect(model->getOptionsModel(), SIGNAL(transactionFeeChanged(qint64)), this, SLOT(coinControlUpdateLabels()));
        ui->frameCoinControl->setVisible(model->getOptionsModel()->getCoinControlFeatures());
        coinControlUpdateLabels();
    }
}

SendCoinsDialog::~SendCoinsDialog()
{
    delete ui;
}

void SendCoinsDialog::on_sendButton_clicked()
{
    if(!model || !model->getOptionsModel()) {
        return;
    }

    QList<SendCoinsRecipient> recipients;
    bool valid = true;

    for(int i = 0; i < ui->entries->count(); ++i) {
        SendCoinsEntry *entry = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(i)->widget());

        if(entry) {
            if(entry->validate()) {
                recipients.append(entry->getValue());
            } else {
                valid = false;
            }
        }
    }

    if(!valid || recipients.isEmpty()) {
        return;
    }

    QString strFunds = tr("using") + " <b>" + tr("anonymous funds") + "</b>";
    QString strFee = "";
    recipients[0].inputType = ONLY_DENOMINATED;

    if(ui->checkUseSandstorm->isChecked()) {
        recipients[0].inputType = ONLY_DENOMINATED;
        strFunds = tr("using") + " <b>" + tr("anonymous funds") + "</b>";
        QString strNearestAmount(
            DarkSilkUnits::formatWithUnit(
                model->getOptionsModel()->getDisplayUnit(), 0.1 * COIN));
        strFee = QString(tr(
            "(sandstorm requires this amount to be rounded up to the nearest %1)."
        ).arg(strNearestAmount));
    } else {
        recipients[0].inputType = ALL_COINS;
        strFunds = tr("using") + " <b>" + tr("any available funds (not recommended)") + "</b>";
    }

    if(ui->checkInstantX->isChecked()) {
        recipients[0].useInstantX = true;
        strFunds += " ";
        strFunds += tr("and InstantX");
    } else {
        recipients[0].useInstantX = false;
    }

    // Format confirmation message
    QStringList formatted;

    foreach(const SendCoinsRecipient &rcp, recipients) {
        formatted.append(tr("<b>%1</b> to %2 (%3)").arg(DarkSilkUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), rcp.amount), Qt::escape(rcp.label), rcp.address));
    }

    fNewRecipientAllowed = false;

    QMessageBox::StandardButton retval = QMessageBox::question(this, tr("Confirm send coins"),
                          tr("Are you sure you want to send %1?").arg(formatted.join(tr(" and "))),
          QMessageBox::Yes|QMessageBox::Cancel,
          QMessageBox::Cancel);

    if(retval != QMessageBox::Yes) {
        fNewRecipientAllowed = true;
        return;
    }

    WalletModel::UnlockContext ctx(model->requestUnlock());

    if(!ctx.isValid()) {
        // Unlock wallet was cancelled
        fNewRecipientAllowed = true;
        return;
    }

    WalletModel::SendCoinsReturn sendstatus;

    if (!model->getOptionsModel() || !model->getOptionsModel()->getCoinControlFeatures()) {
        sendstatus = model->sendCoins(recipients);
    } else {
        sendstatus = model->sendCoins(recipients, CoinControlDialog::coinControl);
    }

    switch(sendstatus.status)
    {
    case WalletModel::InvalidAddress:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("The recipient address is not valid, please recheck."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::InvalidAmount:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("The amount to pay must be larger than 0."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::AmountExceedsBalance:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("The amount exceeds your balance."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::AmountWithFeeExceedsBalance:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("The total exceeds your balance when the %1 transaction fee is included.").
            arg(DarkSilkUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), sendstatus.fee)),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::DuplicateAddress:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("Duplicate address found, can only send to each address once per send operation."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::TransactionCreationFailed:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("Error: Transaction creation failed!"),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::TransactionCommitFailed:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("Error: The transaction was rejected. This might happen if some of the coins in your wallet were already spent, such as if you used a copy of wallet.dat and coins were spent in the copy but not marked as spent here."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::NarrationTooLong:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("Error: Narration is too long."),
            QMessageBox::Ok, QMessageBox::Ok);
        break;
    case WalletModel::Aborted: // User aborted, nothing to do
        break;    
    case WalletModel::AnonymizeOnlyUnlocked:
        QMessageBox::warning(this, tr("Send Coins"),
            tr("Error: The wallet was unlocked only to anonymize coins."),
            QMessageBox::Ok, QMessageBox::Ok);
    case WalletModel::OK:
        accept();
        CoinControlDialog::coinControl->UnSelectAll();
        coinControlUpdateLabels();
        break;
    }

    fNewRecipientAllowed = true;
}

void SendCoinsDialog::clear()
{
    // Remove entries until only one left
    while(ui->entries->count())
    {
        delete ui->entries->takeAt(0)->widget();
    }
    addEntry();

    updateRemoveEnabled();

    ui->sendButton->setDefault(true);
}

void SendCoinsDialog::reject()
{
    clear();
}

void SendCoinsDialog::accept()
{
    clear();
}

SendCoinsEntry *SendCoinsDialog::addEntry()
{
    SendCoinsEntry *entry = new SendCoinsEntry(this);
    entry->setModel(model);
    ui->entries->addWidget(entry);
    connect(entry, SIGNAL(removeEntry(SendCoinsEntry*)), this, SLOT(removeEntry(SendCoinsEntry*)));
    connect(entry, SIGNAL(payAmountChanged()), this, SLOT(coinControlUpdateLabels()));

    updateRemoveEnabled();

    // Focus the field, so that entry can start immediately
    entry->clear();
    entry->setFocus();
    ui->scrollAreaWidgetContents->resize(ui->scrollAreaWidgetContents->sizeHint());
    QCoreApplication::instance()->processEvents();
    QScrollBar* bar = ui->scrollArea->verticalScrollBar();
    if(bar)
        bar->setSliderPosition(bar->maximum());
    return entry;
}

void SendCoinsDialog::updateRemoveEnabled()
{
    // Remove buttons are enabled as soon as there is more than one send-entry
    bool enabled = (ui->entries->count() > 1);
    for(int i = 0; i < ui->entries->count(); ++i)
    {
        SendCoinsEntry *entry = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(i)->widget());
        if(entry)
        {
            entry->setRemoveEnabled(enabled);
        }
    }
    setupTabChain(0);
    coinControlUpdateLabels();
}

void SendCoinsDialog::removeEntry(SendCoinsEntry* entry)
{
    delete entry;
    updateRemoveEnabled();
}

QWidget *SendCoinsDialog::setupTabChain(QWidget *prev)
{
    for(int i = 0; i < ui->entries->count(); ++i)
    {
        SendCoinsEntry *entry = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(i)->widget());
        if(entry)
        {
            prev = entry->setupTabChain(prev);
        }
    }
    QWidget::setTabOrder(prev, ui->addButton);
    QWidget::setTabOrder(ui->addButton, ui->sendButton);
    return ui->sendButton;
}

void SendCoinsDialog::pasteEntry(const SendCoinsRecipient &rv)
{
    if(!fNewRecipientAllowed)
        return;

    SendCoinsEntry *entry = 0;
    // Replace the first entry if it is still unused
    if(ui->entries->count() == 1)
    {
        SendCoinsEntry *first = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(0)->widget());
        if(first->isClear())
        {
            entry = first;
        }
    }
    if(!entry)
    {
        entry = addEntry();
    }

    entry->setValue(rv);
}

bool SendCoinsDialog::handleURI(const QString &uri)
{
    SendCoinsRecipient rv;
    // URI has to be valid
    if (GUIUtil::parseDarkSilkURI(uri, &rv))
    {
        CDarkSilkAddress address(rv.address.toStdString());
        if (!address.IsValid())
            return false;
        pasteEntry(rv);
        return true;
    }

    return false;
}

void SendCoinsDialog::setBalance(const CAmount& balance, const CAmount& stake, const CAmount& unconfirmedBalance, const CAmount& immatureBalance, const CAmount& anonymizedBalance,
    const CAmount& watchBalance, const CAmount& watchStake, const CAmount& watchUnconfirmedBalance, const CAmount& watchImmatureBalance)
{
    Q_UNUSED(stake);
    Q_UNUSED(unconfirmedBalance);
    Q_UNUSED(immatureBalance);
    Q_UNUSED(anonymizedBalance);
    Q_UNUSED(watchBalance);
    Q_UNUSED(watchStake);
    Q_UNUSED(watchUnconfirmedBalance);
    Q_UNUSED(watchImmatureBalance);

    if(model && model->getOptionsModel())
    {
        CAmount bal = 0;
        QSettings settings;
        settings.setValue("bUseSandStorm", ui->checkUseSandstorm->isChecked());
        if(ui->checkUseSandstorm->isChecked()) {
            bal = anonymizedBalance;
        } else {
            bal = balance;
        }

        ui->labelBalance->setText(DarkSilkUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), bal));
    }
}

void SendCoinsDialog::updateDisplayUnit()
{
    TRY_LOCK(cs_main, lockMain);
    if(!lockMain) return;
    setBalance(model->getBalance(), model->getStake(), model->getUnconfirmedBalance(), model->getImmatureBalance(), model->getAnonymizedBalance(),
        model->getWatchBalance(), model->getWatchStake(), model->getWatchUnconfirmedBalance(), model->getWatchImmatureBalance());
    CoinControlDialog::coinControl->useSandStorm = ui->checkUseSandstorm->isChecked();
    coinControlUpdateLabels();
}

void SendCoinsDialog::updateInstantX()
{   
    QSettings settings;
    settings.setValue("bUseInstantX", ui->checkInstantX->isChecked());
    CoinControlDialog::coinControl->useInstantX = ui->checkInstantX->isChecked();
    coinControlUpdateLabels();
}

// Coin Control: copy label "Quantity" to clipboard
void SendCoinsDialog::coinControlClipboardQuantity()
{
    QApplication::clipboard()->setText(ui->labelCoinControlQuantity->text());
}

// Coin Control: copy label "Amount" to clipboard
void SendCoinsDialog::coinControlClipboardAmount()
{
    QApplication::clipboard()->setText(ui->labelCoinControlAmount->text().left(ui->labelCoinControlAmount->text().indexOf(" ")));
}

// Coin Control: copy label "Fee" to clipboard
void SendCoinsDialog::coinControlClipboardFee()
{
    QApplication::clipboard()->setText(ui->labelCoinControlFee->text().left(ui->labelCoinControlFee->text().indexOf(" ")));
}

// Coin Control: copy label "After fee" to clipboard
void SendCoinsDialog::coinControlClipboardAfterFee()
{
    QApplication::clipboard()->setText(ui->labelCoinControlAfterFee->text().left(ui->labelCoinControlAfterFee->text().indexOf(" ")));
}

// Coin Control: copy label "Bytes" to clipboard
void SendCoinsDialog::coinControlClipboardBytes()
{
    QApplication::clipboard()->setText(ui->labelCoinControlBytes->text());
}

// Coin Control: copy label "Priority" to clipboard
void SendCoinsDialog::coinControlClipboardPriority()
{
    QApplication::clipboard()->setText(ui->labelCoinControlPriority->text());
}

// Coin Control: copy label "Dust" to clipboard
void SendCoinsDialog::coinControlClipboardLowOutput()
{
    QApplication::clipboard()->setText(ui->labelCoinControlLowOutput->text());
}

// Coin Control: copy label "Change" to clipboard
void SendCoinsDialog::coinControlClipboardChange()
{
    QApplication::clipboard()->setText(ui->labelCoinControlChange->text().left(ui->labelCoinControlChange->text().indexOf(" ")));
}

// Coin Control: settings menu - coin control enabled/disabled by user
void SendCoinsDialog::coinControlFeatureChanged(bool checked)
{
    ui->frameCoinControl->setVisible(checked);

    if (!checked && model) // coin control features disabled
        CoinControlDialog::coinControl->SetNull();
}

// Coin Control: button inputs -> show actual coin control dialog
void SendCoinsDialog::coinControlButtonClicked()
{
    CoinControlDialog dlg;
    dlg.setModel(model);
    dlg.exec();
    coinControlUpdateLabels();
}

// Coin Control: checkbox custom change address
void SendCoinsDialog::coinControlChangeChecked(int state)
{
    if (model)
    {
        if (state == Qt::Checked)
            CoinControlDialog::coinControl->destChange = CDarkSilkAddress(ui->lineEditCoinControlChange->text().toStdString()).Get();
        else
            CoinControlDialog::coinControl->destChange = CNoDestination();
    }

    ui->lineEditCoinControlChange->setEnabled((state == Qt::Checked));
    ui->labelCoinControlChangeLabel->setEnabled((state == Qt::Checked));
}

// Coin Control: custom change address changed
void SendCoinsDialog::coinControlChangeEdited(const QString & text)
{
    if (model)
    {
        CoinControlDialog::coinControl->destChange = CDarkSilkAddress(text.toStdString()).Get();

        // label for the change address
        ui->labelCoinControlChangeLabel->setStyleSheet("QLabel{color:black;}");
        if (text.isEmpty())
            ui->labelCoinControlChangeLabel->setText("");
        else if (!CDarkSilkAddress(text.toStdString()).IsValid())
        {
            ui->labelCoinControlChangeLabel->setStyleSheet("QLabel{color:red;}");
            ui->labelCoinControlChangeLabel->setText(tr("WARNING: Invalid DarkSilk address"));
        }
        else
        {
            QString associatedLabel = model->getAddressTableModel()->labelForAddress(text);
            if (!associatedLabel.isEmpty())
                ui->labelCoinControlChangeLabel->setText(associatedLabel);
            else
            {
                CPubKey pubkey;
                CKeyID keyid;
                CDarkSilkAddress(text.toStdString()).GetKeyID(keyid);   
                if (model->getPubKey(keyid, pubkey))
                    ui->labelCoinControlChangeLabel->setText(tr("(no label)"));
                else
                {
                    ui->labelCoinControlChangeLabel->setStyleSheet("QLabel{color:red;}");
                    ui->labelCoinControlChangeLabel->setText(tr("WARNING: unknown change address"));
                }
            }
        }
    }
}

// Coin Control: update labels
void SendCoinsDialog::coinControlUpdateLabels()
{
    if (!model || !model->getOptionsModel() || !model->getOptionsModel()->getCoinControlFeatures())
        return;

    // set pay amounts
    CoinControlDialog::payAmounts.clear();
    for(int i = 0; i < ui->entries->count(); ++i)
    {
        SendCoinsEntry *entry = qobject_cast<SendCoinsEntry*>(ui->entries->itemAt(i)->widget());
        if(entry)
            CoinControlDialog::payAmounts.append(entry->getValue().amount);
    }

    ui->checkUseSandstorm->setChecked(CoinControlDialog::coinControl->useSandStorm);

    if (CoinControlDialog::coinControl->HasSelected())
    {
        // actual coin control calculation
        CoinControlDialog::updateLabels(model, this);

        // show coin control stats
        ui->labelCoinControlAutomaticallySelected->hide();
        ui->widgetCoinControl->show();
    }
    else
    {
        // hide coin control stats
        ui->labelCoinControlAutomaticallySelected->show();
        ui->widgetCoinControl->hide();
        ui->labelCoinControlInsuffFunds->hide();
    }
}
